### DISCUSSION TITLE
[General]: Python shell warnings

### DISCUSSION THREAD

**Discussion by cloudchin** [d12frosted/homebrew-emacs-plus#772] [General]
### Issue description

Starting the python interpreter while editing a python file via `C-c C-p`, I get the follow emacs warning:

```
Warning (python): Your ‘python-shell-interpreter’ doesn’t seem to support readline, yet ‘python-shell-completion-native-enable’ was t and "python3" is not part of the ‘python-shell-completion-native-disabled-interpreters’ list.  Native completions have been disabled locally. Consider installing the python package "readline". 
```

The opened python shell includes its own warning:
```
Python 3.13.1 (main, Dec  3 2024, 17:59:52) [Clang 16.0.0 (clang-1600.0.26.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
warning: can't use pyrepl: terminal doesn't have the required clear capability; TERM=dumb
>>> 
```

I found this recent thread: [Python interpreter doesn't seem to support readline error in Emacs on MacOS](https://www.danliden.com/notes/20240709-python-readline.html) which recommends installing `gnureadline` via pip. I try to run this and the command doesn't run throwing a warning about everything being managed by brew. Which I suppose is partially why I look forward to using your formula, I dont' want to deal with dependencies.

I also see that homebrew [switched from gnu readline to libedit](https://github.com/Homebrew/homebrew-core/issues/123842) in python 3.11, and your formula installs python 3.13.

I am not familiar with formula specifications, but is this something that can be fixed?

**Comment by d12frosted** [comment]
Hey @cloudchin 

Emacs+ doesn't depend on Python directly. Some dependencies might require it though.

Keep in mind that macOS runs `.app` files launched via Finder/Spotlight in the login shell, meaning that whatever you have in your interactive shell configurations is not inherited in applications. And macOS comes bundled with some Python version. So you might be hitting this. We have PATH injection in Emacs+ to help with this kind of issues, but it's not ideal and it doesn't seem to be working in latest versions of macOS. You can verify which Python is used by running `M-! which python <return>` (or python3 🤷 ). Maybe you'll need to use something like https://github.com/purcell/exec-path-from-shell. 

I am not using Python myself, so not sure if it makes sense or not, but I think venv should also help to manage versions in specific project. Albeit I am sure it requires some setup though.

In any case, it doesn't seem to be related to Emacs+, and I also see no action items here, so moving this to discussions.

**Comment by cloudchin** [comment]
Howdy.

I read a bit of your notes on PATH considerations, and I boiled it down to "run emacs from shell, not finder" -- which I've done. 

Either way, I don't think that has anything to do with this issue because both the system python and now the python packaged with emacs-plus (which your formula requires, whether indirectly or directly) ships with libedit.

I also don't want to setup a virtual environment in python and add complexity to what should be my 'default' Mac configuration. If there's a way to install a simple package to make it work with the python emacs-plus ships with, great. I'd like to reserve virtual environment configurations for something actually specific to the projects I am working. Hoping someone has the answer!
